File: packages/core/src/factories/CommentFactory.ts
import ts from "typescript";

export namespace CommentFactory {
  export const description = (
    symbol: ts.Symbol,
    includeTags: boolean = false,
  ): string | undefined => {
    const node = symbol.declarations?.[0];
    if (!node) return undefined;

    // FOR LEGACY TS < 5.2
    const [major, minor] = ts.versionMajorMinor.split(".").map(Number) as [
      number,
      number,
    ];
    if (major < 5 || (major === 5 && minor < 1)) {
      const content: string[] = [];
      const main: string = ts.displayPartsToString(
        symbol.getDocumentationComment(undefined),
      );
      if (main.length) {
        content.push(main);
        if (includeTags && symbol.getJsDocTags().length) content.push("");
      }
      if (includeTags)
        for (const tag of symbol.getJsDocTags()) {
          content.push(
            tag.text
              ? `@${tag.name} ${ts.displayPartsToString(tag.text)}`
              : `@${tag.name}`,
          );
        }
      return content.length
        ? content.map((line) => line.split("\r\n").join("\n")).join("\n")
        : undefined;
    }

    // NEW FEATURE OF TS 5.2
    const elements: readonly (ts.JSDoc | ts.JSDocTag)[] =
      ts.getJSDocCommentsAndTags(node);
    if (elements.length === 0) return undefined;

    const content: string[] = [];
    for (const comment of elements) {
      if (ts.isJSDoc(comment)) {
        const parsed: string | undefined = ts.getTextOfJSDocComment(
          comment.comment,
        );
        if (parsed?.length) {
          content.push(parsed);
          if (includeTags && comment.tags?.length) content.push("");
        }
        if (includeTags)
          for (const tag of comment.tags ?? [])
            content.push(parseJSDocTag(tag));
      } else if (includeTags) content.push(parseJSDocTag(comment));
    }
    const output: string = content
      .map((line) => line.split("\r\n").join("\n"))
      .join("\n");
    return output.length ? output : undefined;
  };

  export const merge = (comments: ts.SymbolDisplayPart[]): string =>
    comments
      .map((part) => part.text)
      .map((str) => str.split("\r\n").join("\n"))
      .join("");
}

const parseJSDocTag = (tag: ts.JSDocTag): string => {
  const name: string | undefined = (
    tag as ts.JSDocParameterTag
  ).name?.getText();
  const parsed: string | undefined = ts.getTextOfJSDocComment(tag.comment);
  return [`@${tag.tagName.text}`, name, parsed]
    .filter((str) => !!str?.length)
    .join(" ");
};

File: packages/core/src/factories/CommentFactory.go
package factories

import "strings"

type commentFactory struct{}

var CommentFactory commentFactory

func (commentFactory) Description(symbol CommentSymbol, includeTags ...bool) *string {
	if symbol == nil || len(symbol.Declarations()) == 0 {
		return nil
	}
	withTags := len(includeTags) > 0 && includeTags[0]
	content := make([]string, 0, 1+len(symbol.JSDocTags()))
	main := CommentFactory.Merge(symbol.Documentation())
	if main != "" {
		content = append(content, main)
		if withTags && len(symbol.JSDocTags()) != 0 {
			content = append(content, "")
		}
	}
	if withTags {
		for _, tag := range symbol.JSDocTags() {
			content = append(content, parseJSDocTag(tag))
		}
	}
	if len(content) == 0 {
		return nil
	}
	output := strings.Join(content, "\n")
	return &output
}

func (commentFactory) Merge(comments []SymbolDisplayPart) string {
	var builder strings.Builder
	for _, part := range comments {
		builder.WriteString(normalizeNewline(part.Text))
	}
	return builder.String()
}

func parseJSDocTag(tag JSDocTag) string {
	parts := []string{"@" + tag.Name}
	if tag.Parameter != "" {
		parts = append(parts, tag.Parameter)
	}
	if tag.Text != "" {
		parts = append(parts, normalizeNewline(tag.Text))
	}
	return strings.Join(parts, " ")
}
